pkg/utils/gateway_wrapper.go

```diff
@@ -0,0 +1,129 @@
+func (g *GatewayWrapper) Validate() error {
+
+	// Status.Addresses validation
+	// Compares all addresses against the first address to ensure the same type
+	for _, address := range g.Status.Addresses {
+		if g.IsMultiCluster() == isMultiClusterAddressType(*address.Type) {
+			continue
+		}
+		return fmt.Errorf("gateway is invalid: inconsistent status addresses")
+	}
+	return nil
+}
+
+// GetClusterGatewayAddresses constructs a map from Status.Addresses of underlying Gateway
+// with key being a cluster and value being an address in the cluster.
+// In case of a single-cluster Gateway the key is SingleClusterAddressValue
+func (g *GatewayWrapper) GetClusterGatewayAddresses() map[string][]gatewayv1beta1.GatewayAddress {
+	clusterAddrs := make(map[string][]gatewayv1beta1.GatewayAddress, len(g.Status.Addresses))
+
+	for _, address := range g.Status.Addresses {
+		//Default to Single Cluster (Normal Gateway Status)
+		cluster := SingleClusterAddressValue
+		addressValue := address.Value
+
+		//Check for Multi Cluster (MGC Gateway Status)
+		if g.IsMultiCluster() {
+			tmpCluster, tmpAddress, found := strings.Cut(address.Value, "/")
+			//If this fails something is wrong and the value hasn't been set correctly
+			if found {
+				cluster = tmpCluster
+				addressValue = tmpAddress
+			}
+		}
+
+		if _, ok := clusterAddrs[cluster]; !ok {
+			clusterAddrs[cluster] = []gatewayv1beta1.GatewayAddress{}
+		}
+
+		clusterAddrs[cluster] = append(clusterAddrs[cluster], gatewayv1beta1.GatewayAddress{
+			Type:  address.Type,
+			Value: addressValue,
+		})
+	}
+
+	return clusterAddrs
+}
+
+// ListenerTotalAttachedRoutes returns a count of attached routes from the Status.Listeners for a specified
+// combination of downstreamClusterName and specListener.Name
+func (g *GatewayWrapper) ListenerTotalAttachedRoutes(downstreamClusterName string, specListener gatewayv1beta1.Listener) int {
+	for _, statusListener := range g.Status.Listeners {
+		// for Multi Cluster (MGC Gateway Status)
+		if g.IsMultiCluster() {
+			clusterName, listenerName, found := strings.Cut(string(statusListener.Name), ".")
+			if !found {
+				return 0
+			}
+			if clusterName == downstreamClusterName && listenerName == string(specListener.Name) {
+				return int(statusListener.AttachedRoutes)
+			}
+		}
+		// Single Cluster (Normal Gateway Status)
+		if string(statusListener.Name) == string(specListener.Name) {
+			return int(statusListener.AttachedRoutes)
+		}
+	}
+
+	return 0
+}
+
+// AddressTypeToMultiCluster returns a multi cluster version of the address type
+// and a bool to indicate that provided address has supported type
+func AddressTypeToMultiCluster(address gatewayv1beta1.GatewayAddress) (gatewayv1beta1.AddressType, bool) {
+	if *address.Type == gatewayv1beta1.IPAddressType {
+		return MultiClusterIPAddressType, true
+	} else if *address.Type == gatewayv1beta1.HostnameAddressType {
+		return MultiClusterHostnameAddressType, true
+	}
+	return "", false
```

Comment:
Not sure where this would be used other than where it is already, but you could have this return the original address type (*address.Type) which might be more useful in other circumstances.

```.go
return *address.Type, false
```

Reply:
Thinking about it, if you wanted to keep the signatures of `AddressTypeToMultiCluster` and `AddressTypeToSingleCluster` the same  (I kind of like being able to do that where possible) we could add back the bool to `AddressTypeToSingleCluster` and have both return `*address.Type, false` in the case where we are not converting it. The bool value tells you if it was converted or not, rather being "supported"

Reply:
hmm, i can see an argument for: if we converted an address to the multicluster type we always must be able to convert it back :thinking:  and I can't see how information of was it converted or not will be useful :thinking: 
But in the same time: consistency and for the future changes - changed.